074-connector-offsets-support.md

```diff
@@ -0,0 +1,264 @@
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaConnector resources.
+This proposal suggests introducing a new annotation on the KafkaConnector resource to trigger each operation.
+The operations are:
+* List offsets.
+* Alter offsets.
+* Reset offsets.
+
+The list offsets operation uses a ConfigMap for the output.
+The alter offsets operation uses a ConfigMap for the input.
+
+### Request/Response from the API endpoints
+
+The response body received when using the GET endpoint, and the request body used when calling the PATCH endpoint have identical structures to allow users to easily fetch and then alter offsets.
+The structure of the body is different for source and sink connectors.
+This is because sink connectors use regular Kafka offsets, whereas source connectors determine their own structure for offsets.
+This difference is reflected in the approaches outlined in this proposal.
+
+Source connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        // Connector-defined source partition
+      },
+      "offset": {
+        // Connector-defined source offset
+      }
+    }
+  ]
+}
+```
+
+Sink connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "kafka_topic": // Kafka topic
+        "kafka_partition": // Kafka partition
+      },
+      "offset": {
+        "kafka_offset": // Kafka offset
+      }
+    }
+  ]
+}
+```
+
+### New annotation
+
+All three actions will make use of a new annotation called `strimzi.io/connectorOffsets`.
+The possible values will be `list`, `alter` and `reset`.
+
+### Listing offsets
+
+To list the current offsets the user will annotate the KafkaConnector resource with the new annotation `strimzi.io/connectorOffsets` set to `list`.
+After the annotation is added, on the next reconciliation the operator will fetch the current offsets and create a Kubernetes ConfigMap containing the JSON response.
+If the ConfigMap already exists the operator will replace the data with the updated data.
+The operator will then remove the annotation from the KafkaConnector CR.
+
+If the user wants to see an updated set of offsets they will need to re-annotate the resource.
+
+The name of the ConfigMap the operator creates/updates will be set by the user in the KafkaConnector CR.
+The KafkaConnector CRD will be updated to contain a new property called `listOffsets` that must be set for the `strimzi.io/connectorOffsets=list` annotation to take effect.
+The structure of the `listOffsets` property will be as below, to allow it to be extended in future if required:
+
+```yaml
+listOffsets:
+  toConfigMap:
+    name: my-connector-offsets
```

Comment:
* What if the config map already exists?
* Also, will it have ownerReference attached to the KafkaConnector resource?

Reply:
I stated higher up that we would overwrite the contents if it existed already. For owner references, I would suggest if the operator creates it we add one, but if it's there already we don't add it.

Reply:
Well, at least until server-side-apply is the main way to do things, you will likely overwrite the CM if it already exists. So keeping any owner references is an additional effort. Also, you can have more of them IIRC.

Reply:
Do you think we shouldn't add any owner references then?

Reply:
TBH, I do not have a strong opinion ... that is why I was curious about what the plan is :-D. I guess it has both pros and cons. The positive thing is that it will be cleaned up when you delete the connector and clean up all related things. The negative is that it will be deleted with the connector even if you would want to keep it. 🤔 So I'm really not sure what is the best.